File: data-structures-sorting-algorithms-advanced/benchmark/benchmark.cpp
```cpp
#include "../src/helpers.h"
#include "../src/sort_implementations.h"
#include "../src/main_algorithms.cpp" // Include problem implementations
#include <iostream>
#include <vector>
#include <string>
#include <algorithm> // For std::sort

namespace SortingAlgorithms {

// --- Benchmarking function for sorting algorithms ---
template <typename T, typename SortFunc>
void benchmark_sort_algo(const std::string& name, SortFunc sort_func, int num_elements, int num_iterations) {
    long double total_time = 0;
    std::vector<T> original_data = generate_random_vector(num_elements, 0, num_elements * 10);

    for (int i = 0; i < num_iterations; ++i) {
        std::vector<T> data_to_sort = original_data; // Work on a copy
        Timer timer;
        sort_func(data_to_sort);
        total_time += timer.elapsed_milliseconds();

        // Optional: Verify if sorted correctly (only for a few iterations or smaller sizes)
        // if (!std::is_sorted(data_to_sort.begin(), data_to_sort.end())) {
        //     std::cerr << "Error: " << name << " failed to sort correctly on iteration " << i << std::endl;
        //     return;
        // }
    }
    std::cout << name << " (N=" << num_elements << ", Iter=" << num_iterations << "): "
              << (total_time / num_iterations) << " ms (avg)" << std::endl;
}

// --- Benchmarking function for Kth Largest Element approaches ---
void benchmark_kth_largest(int num_elements, int num_iterations) {
    std::cout << "\n--- Benchmarking Kth Largest Element (N=" << num_elements << ", Iter=" << num_iterations << ") ---" << std::endl;

    std::vector<int> original_data = generate_random_vector(num_elements, 0, num_elements * 10);
    int k = num_elements / 2; // Find median-like element

    // QuickSelect
    long double qs_total_time = 0;
    for (int i = 0; i < num_iterations; ++i) {
        std::vector<int> data_copy = original_data;
        Timer timer;
        findKthLargest_QuickSelect(data_copy, k);
        qs_total_time += timer.elapsed_milliseconds();
    }
    std::cout << "QuickSelect: " << (qs_total_time / num_iterations) << " ms (avg)" << std::endl;

    // Min-Heap
    long double heap_total_time = 0;
    for (int i = 0; i < num_iterations; ++i) {
        std::vector<int> data_copy = original_data; // MinHeap takes const vector
        Timer timer;
        findKthLargest_MinHeap(data_copy, k);
        heap_total_time += timer.elapsed_milliseconds();
    }
    std::cout << "Min-Heap:    " << (heap_total_time / num_iterations) << " ms (avg)" << std::endl;

    // Sort (Brute Force)
    long double sort_total_time = 0;
    for (int i = 0; i < num_iterations; ++i) {
        std::vector<int> data_copy = original_data;
        Timer timer;
        findKthLargest_Sort(data_copy, k);
        sort_total_time += timer.elapsed_milliseconds();
    }
    std::cout << "Sort & Pick: " << (sort_total_time / num_iterations) << " ms (avg)" << std::endl;
}


// --- Benchmarking function for Sort Colors approaches ---
void benchmark_sort_colors(int num_elements, int num_iterations) {
    std::cout << "\n--- Benchmarking Sort Colors (N=" << num_elements << ", Iter=" << num_iterations << ") ---" << std::endl;

    // Generate data with 0s, 1s, 2s
    std::vector<int> original_data(num_elements);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distrib(0, 2);
    for (int i = 0; i < num_elements; ++i) {
        original_data[i] = distrib(gen);
    }

    // Dutch National Flag
    long double dutch_total_time = 0;
    for (int i = 0; i < num_iterations; ++i) {
        std::vector<int> data_copy = original_data;
        Timer timer;
        sortColors_DutchNationalFlag(data_copy);
        dutch_total_time += timer.elapsed_milliseconds();
    }
    std::cout << "Dutch National Flag: " << (dutch_total_time / num_iterations) << " ms (avg)" << std::endl;

    // Counting Sort
    long double counting_total_time = 0;
    for (int i = 0; i < num_iterations; ++i) {
        std::vector<int> data_copy = original_data;
        Timer timer;
        sortColors_CountingSort(data_copy);
        counting_total_time += timer.elapsed_milliseconds();
    }
    std::cout << "Counting Sort:       " << (counting_total_time / num_iterations) << " ms (avg)" << std::endl;
}


// --- Benchmarking function for Meeting Rooms II approaches ---
void benchmark_min_meeting_rooms(int num_meetings, int num_iterations) {
    std::cout << "\n--- Benchmarking Meeting Rooms II (N=" << num_meetings << ", Iter=" << num_iterations << ") ---" << std::endl;

    std::vector<Interval> original_meetings(num_meetings);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distrib_start(0, num_meetings * 5); // Start times up to 5*N
    std::uniform_int_distribution<> distrib_duration(1, 10);           // Duration 1 to 10

    for (int i = 0; i < num_meetings; ++i) {
        int start = distrib_start(gen);
        int end = start + distrib_duration(gen);
        original_meetings[i] = {start, end};
    }

    // Min-Heap approach
    long double heap_total_time = 0;
    for (int i = 0; i < num_iterations; ++i) {
        std::vector<Interval> meetings_copy = original_meetings;
        Timer timer;
        minMeetingRooms_Heap(meetings_copy);
        heap_total_time += timer.elapsed_milliseconds();
    }
    std::cout << "Min-Heap Approach: " << (heap_total_time / num_iterations) << " ms (avg)" << std::endl;

    // Sweep Line approach
    long double sweep_total_time = 0;
    for (int i = 0; i < num_iterations; ++i) {
        std::vector<Interval> meetings_copy = original_meetings;
        Timer timer;
        minMeetingRooms_SweepLine(meetings_copy);
        sweep_total_time += timer.elapsed_milliseconds();
    }
    std::cout << "Sweep Line Approach: " << (sweep_total_time / num_iterations) << " ms (avg)" << std::endl;
}


int main() {
    std::cout << "--- Sorting Algorithms Benchmarks ---" << std::endl;

    int N_small = 1000;
    int N_medium = 10000;
    int N_large = 50000;
    int N_very_large = 100000;

    int iterations_small = 1000;
    int iterations_medium = 100;
    int iterations_large = 10;
    int iterations_very_large = 5;


    // --- Benchmark Core Sorting Algorithms ---
    std::cout << "\n--- Core Sorting Algorithms (N=" << N_medium << ", Iter=" << iterations_medium << ") ---" << std::endl;
    benchmark_sort_algo<int>("std::sort", [](std::vector<int>& arr){ std::sort(arr.begin(), arr.end()); }, N_medium, iterations_medium);
    benchmark_sort_algo<int>("Quick Sort", [](std::vector<int>& arr){ SortingAlgorithms::quickSort(arr); }, N_medium, iterations_medium);
    benchmark_sort_algo<int>("Merge Sort", [](std::vector<int>& arr){ SortingAlgorithms::mergeSort(arr); }, N_medium, iterations_medium);
    benchmark_sort_algo<int>("Heap Sort", [](std::vector<int>& arr){ SortingAlgorithms::heapSort(arr); }, N_medium, iterations_medium);
    // These N^2 sorts are too slow for N=10000, reducing N
    std::cout << "\n--- Core Sorting Algorithms (N=" << N_small << ", Iter=" << iterations_small << ") ---" << std::endl;
    benchmark_sort_algo<int>("Insertion Sort", [](std::vector<int>& arr){ SortingAlgorithms::insertionSort(arr); }, N_small, iterations_small);
    benchmark_sort_algo<int>("Selection Sort", [](std::vector<int>& arr){ SortingAlgorithms::selectionSort(arr); }, N_small, iterations_small);
    benchmark_sort_algo<int>("Bubble Sort", [](std::vector<int>& arr){ SortingAlgorithms::bubbleSort(arr); }, N_small, iterations_small);

    // --- Benchmark Main Algorithm Problems ---
    benchmark_kth_largest(N_large, iterations_large);
    benchmark_sort_colors(N_very_large, iterations_very_large);
    benchmark_min_meeting_rooms(N_large, iterations_large);

    std::cout << "\n--- Benchmarks Complete ---" << std::endl;

    return 0;
}

} // namespace SortingAlgorithms

int main() {
    return SortingAlgorithms::main(); // Call the benchmark's main
}
```

File: data-structures-sorting-algorithms-advanced/tests/benchmark.js
```javascript
/**
 * @fileoverview Performance benchmarking for sorting algorithms and problem solutions.
 * This script compares the execution time of different sorting algorithms and
 * multiple approaches for selected interview problems on various input sizes.
 */

const { generateRandomArray, isSorted } = require('../src/utils/arrayUtils');

// Sorting Algorithms
const bubbleSort = require('../src/algorithms/bubbleSort');
const selectionSort = require('../src/algorithms/selectionSort');
const insertionSort = require('../src/algorithms/insertionSort');
const mergeSort = require('../src/algorithms/mergeSort');
const quickSort = require('../src/algorithms/quickSort');
const heapSort = require('../src/algorithms/heapSort');

// Problem Solutions
const {
  findKthLargest_sort,
  findKthLargest_minHeap,
  findKthLargest_quickSelect,
} = require('../src/problems/problem1_kthLargestElement');

const mergeIntervals = require('../src/problems/problem2_mergeIntervals');

const {
  sortColors_twoPass,
  sortColors_onePass,
} = require('../src/problems/problem3_dutchNationalFlag');


// --- Benchmarking Utilities ---

/**
 * Runs a function and measures its execution time.
 * @param {Function} func The function to benchmark.
 * @param {Array<any>} args Arguments to pass to the function.
 * @returns {number} The execution time in milliseconds.
 */
function measureTime(func, args) {
  const start = process.hrtime.bigint();
  func(...args);
  const end = process.hrtime.bigint();
  return Number(end - start) / 1_000_000; // Convert nanoseconds to milliseconds
}

/**
 * Executes a benchmark for a given function across different array sizes.
 * @param {string} name Name of the algorithm/function.
 * @param {Function} func The function to benchmark.
 * @param {Array<number>} inputSizes An array of input sizes to test.
 * @param {Function} arrayGenerator A function that takes size, min, max and returns an array.
 * @param {Function} validator An optional function to validate the output (e.g., isSorted).
 * @param {boolean} copyInput If true, pass a copy of the array to the function.
 * @param {Array<any>} additionalArgs Additional arguments to pass to the function besides the array.
 */
function runBenchmark(name, func, inputSizes, arrayGenerator, validator = null, copyInput = true, additionalArgs = []) {
  console.log(`\n--- Benchmarking: ${name} ---`);
  inputSizes.forEach(size => {
    let arr = arrayGenerator(size, 0, size * 2); // Generate unique arrays for each size
    let result = null;
    let time = 0;
    try {
      if (copyInput) {
        // Deep copy for objects in problems
        const inputForFunc = JSON.parse(JSON.stringify(arr));
        time = measureTime(func, [inputForFunc, ...additionalArgs]);
        result = inputForFunc; // For in-place modifications
        // If the function returns a new array (like mergeSort), capture that.
        // For general purpose, let's assume if it returns something, it's the result.
        // Otherwise, the input is modified.
        const returnedResult = func(JSON.parse(JSON.stringify(arr)), ...additionalArgs);
        if (returnedResult !== undefined && returnedResult !== inputForFunc) {
            result = returnedResult;
        }
      } else {
        time = measureTime(func, [arr, ...additionalArgs]);
        result = arr;
      }
    } catch (e) {
      console.error(`  Size ${size}: Error during execution: ${e.message}`);
      return;
    }

    let validationStatus = 'N/A';
    if (validator) {
      try {
        validationStatus = validator(result, ...additionalArgs) ? 'PASSED' : 'FAILED';
      } catch (e) {
        validationStatus = `ERROR: ${e.message}`;
      }
    }
    console.log(`  Size ${size}: ${time.toFixed(3).padStart(8, ' ')} ms (Validation: ${validationStatus})`);
  });
}

// --- Benchmark Configurations ---

const sizes = [100, 1000, 5000, 10000, 50000]; // Array sizes for sorting algorithms

// Sorting algorithms benchmark
const sortingAlgorithmsToBenchmark = [
  { name: 'Bubble Sort', func: bubbleSort, copyInput: true },
  { name: 'Selection Sort', func: selectionSort, copyInput: true },
  { name: 'Insertion Sort', func: insertionSort, copyInput: true },
  { name: 'Merge Sort', func: mergeSort, copyInput: true }, // Merge sort returns new array
  { name: 'Quick Sort', func: quickSort, copyInput: true },
  { name: 'Heap Sort', func: heapSort, copyInput: true },
];

// Problem 1: Kth Largest Element
const kthLargestInputSizes = [100, 1000, 10000, 100000, 500000];
const kthLargestK = (size) => Math.floor(size / 2); // Example: always find the median-ish element

// This validator verifies if the found element is indeed the k-th largest by sorting
const validateKthLargest = (result, originalNums, k) => {
  const sorted = [...originalNums].sort((a, b) => a - b);
  return result === sorted[originalNums.length - k];
};

const kthLargestApproaches = [
  { name: 'Kth Largest (Sort)', func: findKthLargest_sort, validator: validateKthLargest, additionalArgs: [kthLargestK] },
  { name: 'Kth Largest (Min-Heap)', func: findKthLargest_minHeap, validator: validateKthLargest, additionalArgs: [kthLargestK] },
  { name: 'Kth Largest (QuickSelect)', func: findKthLargest_quickSelect, validator: validateKthLargest, additionalArgs: [kthLargestK] },
];

// Problem 2: Merge Intervals
const mergeIntervalsInputSizes = [100, 1000, 5000, 10000];
// Intervals generator: generates non-overlapping and overlapping intervals
const generateIntervals = (size) => {
  const intervals = [];
  let currentStart = 0;
  for (let i = 0; i < size; i++) {
    // Random interval length between 1 and 100
    const length = Math.floor(Math.random() * 100) + 1;
    // Add some overlap potential, or space out
    const gap = Math.floor(Math.random() * 50); // Gap or overlap value
    const start = currentStart + gap;
    const end = start + length;
    intervals.push([start, end]);
    currentStart = start + length + Math.floor(Math.random() * 20); // Move currentStart forward
  }
  return intervals;
};

const validateMergedIntervals = (merged, originalIntervals) => {
    // For robust validation, you'd need a more complex check, e.g.,
    // 1. Ensure all original points are covered.
    // 2. Ensure no overlaps in merged.
    // 3. Ensure merged is sorted.
    // For simplicity here, we'll just check if it's sorted and no trivial overlaps.
    for (let i = 0; i < merged.length - 1; i++) {
        if (merged[i][1] >= merged[i+1][0]) {
            return false; // Found overlap
        }
    }
    return true; // Simplistic validation
};

const mergeIntervalsApproach = [
  { name: 'Merge Intervals', func: mergeIntervals, generator: generateIntervals, validator: validateMergedIntervals, copyInput: true },
];

// Problem 3: Sort Colors (Dutch National Flag)
const sortColorsInputSizes = [1000, 10000, 100000, 1000000];
// Generator for 0, 1, 2
const generateColorsArray = (size) => generateRandomArray(size, 0, 2);

const validateSortedColors = (arr) => {
  // Check if all 0s come first, then 1s, then 2s
  let first1 = -1;
  let first2 = -1;
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === 1 && first1 === -1) first1 = i;
    if (arr[i] === 2 && first2 === -1) first2 = i;

    if (first1 !== -1 && arr[i] === 0) return false; // 0 after 1s started
    if (first2 !== -1 && (arr[i] === 0 || arr[i] === 1)) return false; // 0 or 1 after 2s started
  }
  return true;
};

const sortColorsApproaches = [
  { name: 'Sort Colors (Two-Pass)', func: sortColors_twoPass, validator: validateSortedColors, copyInput: true },
  { name: 'Sort Colors (One-Pass DNF)', func: sortColors_onePass, validator: validateSortedColors, copyInput: true },
];

// --- Main Benchmark Runner ---
function runAllBenchmarks() {
  console.log('Starting Benchmarks...\n');
  console.log('Note: Times are in milliseconds. "N/A" for validation means no specific validator was provided or it threw an error.');

  // Benchmarking Sorting Algorithms
  sortingAlgorithmsToBenchmark.forEach(algo => {
    runBenchmark(
      algo.name,
      algo.func,
      sizes,
      generateRandomArray, // Random arrays are typical for avg case
      isSorted,             // Validate that the output is sorted
      algo.copyInput
    );
  });

  // Benchmarking Kth Largest Element approaches
  kthLargestApproaches.forEach(approach => {
    runBenchmark(
      approach.name,
      approach.func,
      kthLargestInputSizes,
      (size) => generateRandomArray(size, 0, size * 2), // Array generator
      (result, arrCopy) => validateKthLargest(result, arrCopy, approach.additionalArgs[0](arrCopy.length)), // Validator needs original array and k
      true, // Always copy input for problems
      [kthLargestK] // Pass k as an additional argument factory
    );
  });


  // Benchmarking Merge Intervals
  mergeIntervalsApproach.forEach(approach => {
    runBenchmark(
      approach.name,
      approach.func,
      mergeIntervalsInputSizes,
      approach.generator, // Use custom generator for intervals
      validateMergedIntervals, // Validate merged intervals
      approach.copyInput
    );
  });

  // Benchmarking Sort Colors
  sortColorsApproaches.forEach(approach => {
    runBenchmark(
      approach.name,
      approach.func,
      sortColorsInputSizes,
      generateColorsArray, // Use custom generator for colors (0,1,2)
      validateSortedColors, // Validate sorted colors
      approach.copyInput
    );
  });

  console.log('\nBenchmarks Finished.');
}

// Run all benchmarks
runAllBenchmarks();
```